vkubelet/pod.go

```diff
@@ -217,24 +221,65 @@ func (s *Server) updatePodStatuses(ctx context.Context) {
 	span.AddAttributes(trace.Int64Attribute("nPods", int64(len(pods))))
 
 	for _, pod := range pods {
-		if pod.Status.Phase == corev1.PodSucceeded ||
-			pod.Status.Phase == corev1.PodFailed ||
-			pod.Status.Reason == podStatusReasonProviderFailed {
-			continue
+		if err := s.updatePodStatus(ctx, pod); err != nil {
+			logger := log.G(ctx).WithField("pod", pod.GetName()).WithField("namespace", pod.GetNamespace()).WithField("status", pod.Status.Phase).WithField("reason", pod.Status.Reason)
+			logger.Error(err)
 		}
+	}
+}
 
-		status, err := s.provider.GetPodStatus(ctx, pod.Namespace, pod.Name)
-		if err != nil {
-			log.G(ctx).WithField("pod", pod.GetName()).WithField("namespace", pod.GetNamespace()).Error("Error retrieving pod status")
-			return
-		}
+func (s *Server) updatePodStatus(ctx context.Context, pod *corev1.Pod) error {
+	ctx, span := trace.StartSpan(ctx, "updatePodStatus")
+	defer span.End()
+	addPodAttributes(span, pod)
+
+	if pod.Status.Phase == corev1.PodSucceeded ||
+		pod.Status.Phase == corev1.PodFailed ||
+		pod.Status.Reason == podStatusReasonProviderFailed {
+		return nil
+	}
 
-		// Update the pod's status
-		if status != nil {
-			pod.Status = *status
-			s.k8sClient.CoreV1().Pods(pod.Namespace).UpdateStatus(pod)
+	status, err := s.provider.GetPodStatus(ctx, pod.Namespace, pod.Name)
+	if err != nil {
+		span.SetStatus(ocstatus.FromError(err))
+		return pkgerrors.Wrap(err, "error retreiving pod status")
+	}
+
+	// Update the pod's status
+	if status != nil {
+		pod.Status = *status
+	} else {
+		// Only change the status when the pod was already up
+		// Only doing so when the pod was successfully running makes sure we don't run into race conditions during pod creation.
+		if pod.Status.Phase == corev1.PodRunning {
```

Comment:
pod in Pending state might also hit this issue right? since the Succeeded and Failed phases are handled in L236, shall we remove this check?

Reply:
Not putting this check was *really* problematic.

Reply:
Maybe I can check "Reason" and make sure it's not in "Creating".

Reply:
The reason "Creating" is only used in ACI. Add this check might break the other providers.

Reply:
Can we check the pod's creationTimestamp and set a timeout threshold?

Reply:
Can you explain what you mean here?

Reply:
if pod.Status.Phase == corev1.PodRunning || pod.ObjectMeta.CreationTimestamp.Add(time.Minute) < time.Now() {
...
}

Reply:
👍 

Updated